src/librustc_macros/src/symbols.rs

```diff
@@ -106,18 +128,31 @@ pub fn symbols(input: TokenStream) -> TokenStream {
     };
 
     // Generate the listed keywords.
-    for keyword in &input.keywords.0 {
-        let name = &keyword.name;
-        let value = &keyword.value;
-        check_dup(&value.value(), &mut errors);
-        prefill_stream.extend(quote! {
-            #value,
-        });
-        keyword_stream.extend(quote! {
-            #[allow(non_upper_case_globals)]
-            pub const #name: Symbol = Symbol::new(#counter);
-        });
-        counter += 1;
+    for (class, keywords) in &input.keywords.0 {
+        let mut class_stream = quote! {};
+        for keyword in keywords {
+            let name = &keyword.name;
+            let value = &keyword.value;
+            check_dup(&value.value(), &mut errors);
+            prefill_stream.extend(quote! {
+                #value,
+            });
+            keyword_stream.extend(quote! {
+                #[allow(non_upper_case_globals)]
+                pub const #name: Symbol = Symbol::new(#counter);
+            });
+            class_stream.extend(quote! {
+                kw::#name,
+            });
+            counter += 1;
+        }
+        if let Some(class) = class {
+            keyword_class_stream.extend(quote! {
+                fn #class(self) -> bool {
+                    [#class_stream].contains(&self)
```

Comment:
It does seem a bit unfortunate to lose the O(1) range-checking here, but I guess this is probably not too hot.

Maybe LLVM takes care of things for us, too, though I somewhat doubt it.

Reply:
This PR is supposed to land only together with https://github.com/rust-lang/rust/pull/74554, which also regresses performance due to inability to do range checks, but improves it more due to inlining.